include/services/internal/hash_table.h

```diff
@@ -0,0 +1,136 @@
+/* file: hash_table.h */
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __DAAL_ONEAPI_INTERNAL_HASH_TABLE_H__
+#define __DAAL_ONEAPI_INTERNAL_HASH_TABLE_H__
+
+#include "services/daal_string.h"
+#include "services/internal/error_handling_helpers.h"
+#include "services/error_indexes.h"
+
+#include <iostream>
+
+
+namespace daal
+{
+namespace services
+{
+namespace internal
+{
+
+template<class T, size_t size>
+class HashTable
+{
+public:
+    HashTable()
+    {
+        for (size_t i = 0; i < size; ++i)
+            _arr[i] = nullptr;
+    }
+
+    ~HashTable()
+    {
+        for (size_t i = 0; i < size; ++i)
+        {
+            delete _arr[i];
+            _arr[i] = nullptr;
+        }
+    }
+
+    bool contain(const services::String &key, services::Status &status)
+    {
+        size_t id = get_unique_hash(key, status);
+        if (!status.ok())
+            return false;
+        return _arr[id] != nullptr;
+    }
+
+    void add(const services::String &key, const services::SharedPtr<T> obj, services::Status &status)
+    {
+        size_t id = get_unique_hash(key, status);
+        if (!status.ok())
+            return;
+        if (_arr[id] == nullptr)
```

Comment:
`get_unique_hash` returns id with _arr[id] exactly equal to `nullptr` or returns error status. Please remove this if-else, you need only the body under `if` statement